test/parallel/test-fs-rm.js

```diff
@@ -287,74 +286,154 @@ function removeAsync(dir) {
+          } catch {
+          }
+
+          if (!isValidState(fs.existsSync(filePath), err)) {
+            throw err;
+          }
+        }));
       }
     }
 
+    // Check endless recursion.
+    // https://github.com/nodejs/node/issues/34580
     {
-      // Check endless recursion.
-      // https://github.com/nodejs/node/issues/34580
-      const dirname = nextDirPath();
-      fs.mkdirSync(dirname, { recursive: true });
-      const root = fs.mkdtempSync(path.join(dirname, 'fs-'));
-      const middle = path.join(root, 'middle');
-      fs.mkdirSync(middle);
-      fs.mkdirSync(path.join(middle, 'leaf')); // Make `middle` non-empty
-      try {
-        const code = makeDirectoryReadOnly(middle, 0o555);
+      function isValidState(exists, err) {
+        // On Windows, we are not allowed to delete a read-only folder yet.
+        // TODO(RaisinTen): Remove Windows special-casing if this lands:
+        // https://github.com/libuv/libuv/pull/3193
+        return common.isWindows ?
+          (exists === true && err?.code === 'EPERM') :
+          (exists === true && err?.code === 'EACCES');
+      }
+
+      {
+        const dirname = nextDirPath();
+        fs.mkdirSync(dirname, { recursive: true });
+        const root = fs.mkdtempSync(path.join(dirname, 'fs-'));
+        const middle = path.join(root, 'middle');
+        const leaf = path.join(middle, 'leaf');
+
+        fs.mkdirSync(middle);
+        fs.mkdirSync(leaf);
+
+        fs.chmodSync(leaf, 0o555);
+        fs.chmodSync(middle, 0o555);
+
+        let err = null;
+
         try {
-          assert.throws(() => {
-            fs.rmSync(root, { recursive: true });
-          }, {
-            code,
-            name: 'Error',
-          });
-        } catch (err) {
-          // Only fail the test if the folder was not deleted.
-          // as in some cases rmSync succesfully deletes read-only folders.
-          if (fs.existsSync(root)) {
+          fs.rmSync(root, { recursive: true });
+        } catch (_err) {
+          err = _err;
+        }
+
+        try {
+          fs.chmodSync(middle, 0o777);
+        } catch {
+        }
+
+        try {
+          fs.chmodSync(leaf, 0o777);
+        } catch {
+        }
+
+        if (!isValidState(fs.existsSync(root), err)) {
+          throw err;
+        }
+      }
+
+      {
+        const dirname = nextDirPath();
+        fs.mkdirSync(dirname, { recursive: true });
+        const root = fs.mkdtempSync(path.join(dirname, 'fs-'));
+        const middle = path.join(root, 'middle');
+        const leaf = path.join(middle, 'leaf');
+
+        fs.mkdirSync(middle);
+        fs.mkdirSync(leaf);
+
+        fs.chmodSync(leaf, 0o555);
+        fs.chmodSync(middle, 0o555);
+
+        fs.rm(root, { recursive: true }, common.mustCall((err) => {
+          try {
+            fs.chmodSync(middle, 0o777);
+          } catch {
+          }
+
+          try {
+            fs.chmodSync(leaf, 0o777);
+          } catch {
+          }
```

Comment:
nit: keeping those in one line to make it clearer that those revert the action of lines 419 and 420
```suggestion
        try { fs.chmodSync(middle, 0o777); } catch {}
        try { fs.chmodSync(leaf, 0o777); } catch {}
```